vsphere_plugin_common/__init__.py

```diff
@@ -596,139 +765,527 @@ def get_server_list(self):
+
+        minimum_used = minimum_disk + vm_memory
+        maximum_used = minimum_used + maximum_disk
+
+        if free_space - minimum_used < 0:
+            return None
+        else:
+            return free_space - maximum_used
+
+    def recurse_resource_pools(self, resource_pool):
+        """
+            Recursively get all child resource pools given a resource pool.
+            Return a list of all resource pools found.
+        """
+        resource_pool_names = []
+        for pool in resource_pool.resourcePool:
+            resource_pool_names.append(pool)
+            resource_pool_names.extend(self.recurse_resource_pools(pool))
+        return resource_pool_names
+
+    def get_host_networks(self, host):
+        """
+            Get all networks attached to this host.
+            Returns a list of dicts in the form:
+            {
+                'name': <name of network>,
+                'switch_distributed': <whether net is distributed>,
+            }
+        """
+        nets = [
+            {
+                'name': net.name,
+                'switch_distributed': self._port_group_is_distributed(net),
+            }
+            for net in host.network
+        ]
+        return nets
+
+    def get_host_resource_pools(self, host):
+        """
+            Get all resource pools available on this host.
+            This will work for hosts inside and outside clusters.
+            A list of resource pools will be returned, e.g.
+            ['Resources', 'myresourcepool', 'anotherone']
+        """
+        base_resource_pool = host.parent.resourcePool
+        resource_pools = [base_resource_pool]
+        child_resource_pools = self.recurse_resource_pools(base_resource_pool)
+        resource_pools.extend(child_resource_pools)
+        return resource_pools
+
+    def get_clusters(self):
+        """
+            Get a list of all clusters.
+        """
+        return self.get_obj_list([vim.ClusterComputeResource])
+
+    def get_datastores(self):
+        """
+            Get a list of all datastores.
+        """
+        return self.get_obj_list([vim.Datastore])
+
+    def get_host_cluster_membership(self, host):
+        """
+            Return the name of the cluster this host is part of,
+            or None if it is not part of a cluster.
+        """
+        if isinstance(host.parent, vim.ClusterComputeResource):
+            return host.parent.name
+        else:
+            return None
+
+    def host_is_usable(self, host):
+        """
+            Return True if this host is usable for deployments,
+            based on its health.
+            Return False otherwise.
+        """
+        if host.overallStatus in (
+            vim.ManagedEntity.Status.green,
+            vim.ManagedEntity.Status.yellow,
+        ) and host.summary.runtime.connectionState == 'connected':
+            # TODO: Check license state (will be yellow for bad license)
```

Comment:
I wasn't able to before because we didn't have something with a broken license. However... I may be able to test it when I deal with the new vcenter deployment today/tomorrow (depending on connectivity).